[PT-D][FSDP] Implement _clip_grad_norm_ for FSDP

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73405

Implement the `_clip_grad_norm_` for FSDP, issue: https://github.com/pytorch/pytorch/issues/72548

Differential Revision: [D34230605](https://our.internmc.facebook.com/intern/diff/D34230605/)

Commit:
[PT-D][FSDP] Implement _clip_grad_norm_ for FSDP

Implement the `_clip_grad_norm_` for FSDP, issue: https://github.com/pytorch/pytorch/issues/72548

Differential Revision: [D34230605](https://our.internmc.facebook.com/intern/diff/D34230605/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_clip_grad_norm.py:
```diff
@@ -0,0 +1,71 @@
+        input = torch.rand(8, 3)
+        norm_type = 2.5
+        model.to(self.rank)
+        model = FSDP(model)
```
Let's test nested models in follow up iterations as well.

rohan-varma commented on test/distributed/fsdp/test_fsdp_clip_grad_norm.py:
```diff
@@ -0,0 +1,71 @@
+        for p in model.parameters():
+            local_norm = torch.linalg.norm(p.grad, norm_type, dtype=torch.float32)
+            total_norm = local_norm ** norm_type
+            dist.all_reduce(total_norm, group=group)
```
It'll by default use the default PG, is there a reason we're passing it in here?

fduwjj replied:
Yep, makes sense. This is not needed indeed.

rohan-varma commented on test/distributed/fsdp/test_fsdp_clip_grad_norm.py:
```diff
@@ -0,0 +1,71 @@
+        for p in model.parameters():
+            local_norm = torch.linalg.norm(p.grad, norm_type, dtype=torch.float32)
+            total_norm = local_norm ** norm_type
+            dist.all_reduce(total_norm, group=group)
```
same comment as L57

rohan-varma replied:
separate 62-65 into a helper function?

rohan-varma commented on test/distributed/fsdp/test_fsdp_clip_grad_norm.py:
```diff
@@ -0,0 +1,71 @@
+        model.to(self.rank)
+        model = FSDP(model)
+        self.assertTrue(len(input) >= self.world_size)
+        in_data = torch.Tensor(input[self.rank]).to(self.rank)
```
nit:
`in_data = torch.Tensor(input[self.rank], device=self.rank)` is a bit more concise.

rohan-varma commented on test/distributed/fsdp/test_fsdp_clip_grad_norm.py:
```diff
@@ -0,0 +1,71 @@
+        self.assertTrue(len(input) >= self.world_size)
+        in_data = torch.Tensor(input[self.rank]).to(self.rank)
+        out = model(in_data)
+        out.float().sum().backward()
```
why is `.float()` needed?

fduwjj replied:
I saw FairScale use it. Maybe some models' output is not float32?

rohan-varma commented on test/distributed/fsdp/test_fsdp_clip_grad_norm.py:
```diff
@@ -0,0 +1,71 @@
+        model = self.SimpleNN([[3, 15], [15, 17]])
+        group = dist.distributed_c10d._get_default_group()
+        input = torch.rand(8, 3)
+        norm_type = 2.5
```
why are we testing 2.5 and should we test other norms as well?

fduwjj replied:
yes, I will make this more generic.

rohan-varma commented on test/distributed/fsdp/test_fsdp_clip_grad_norm.py:
```diff
@@ -0,0 +1,71 @@
+        in_data = torch.Tensor(input[self.rank]).to(self.rank)
+        out = model(in_data)
+        out.float().sum().backward()
+        fsdp_modules_list = []
```
unused?

rohan-varma commented on test/distributed/fsdp/test_fsdp_clip_grad_norm.py:
```diff
@@ -0,0 +1,71 @@
+            total_norm = local_norm ** norm_type
+            dist.all_reduce(total_norm, group=group)
+            total_norms_after_clip.append(total_norm ** (1.0 / norm_type))
+        self.assertTrue(all(norm <= norm_cap for norm in total_norms_after_clip))
```
can you also assert that this is not the case prior to clipping just to make sure the gradient clipping did something?

fduwjj replied:
sure.

rohan-varma commented on test/distributed/fsdp/test_fsdp_clip_grad_norm.py:
```diff
@@ -0,0 +1,71 @@
+            local_norm = torch.linalg.norm(p.grad, norm_type, dtype=torch.float32)
+            total_norm = local_norm ** norm_type
+            dist.all_reduce(total_norm, group=group)
+            total_norms_after_clip.append(total_norm ** (1.0 / norm_type))
```
This test looks good to me overall. A couple more suggestions:

1. Check the corresponding tests in FairScale repo and port those over if they test something different than this.
2. Verify parity using `nn.utils.clip_grad_norm` on a regular `nn.Module`. Basically, run the FSDP model + FSDP clipping, then regular model + regular clipping, and ensure they are equal. You'll probably have to use `summon_full_parameters()` to verify this.

rohan-varma commented on test/distributed/fsdp/test_fsdp_clip_grad_norm.py:
```diff
@@ -0,0 +1,71 @@
+            return self.fc2(self.gelu(self.fc1(inp)))
+
+    @skip_if_lt_x_gpu(2)
+    def test_one_iteration(self):
```
let's parametrize over many norm types, including the inf norm.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1298,3 +1352,25 @@ def _alloc_storage(data: torch.Tensor, size: torch.Size) -> None:
+        Total norm of the parameters (viewed as a single vector).
+    """
+    if isinstance(parameters, torch.Tensor):
+        parameters = [parameters]
```
`parameters` is typed as a List[Parameter], are there cases we can just get a single tensor?

fduwjj replied:
That's a good point, this is also copied from FairScale. Let me remove this check.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1298,3 +1352,25 @@ def _alloc_storage(data: torch.Tensor, size: torch.Size) -> None:
+    """
+    if isinstance(parameters, torch.Tensor):
+        parameters = [parameters]
+    parameters = list(filter(lambda par: par.grad is not None, parameters))
```
`params_with_grad = [p for p in parameters if p.grad is not None]` might be a bit simpler.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1298,3 +1352,25 @@ def _alloc_storage(data: torch.Tensor, size: torch.Size) -> None:
+        local_norm = torch.linalg.norm(
+            torch.stack([torch.linalg.norm(par.grad.detach(), p, dtype=torch.float32) for par in parameters]), p
+        ).to(dtype=parameters[0].dtype)  # type: ignore
+    return local_norm
```
probably worth having a basic unittest for `_calc_grad_norm`. Is there also no other helper/utility that will return the gradient norm of a given set of parameters? There doesn't appear to be any FSDP specific logic here so we could separate it out into a more generally accessible folder in the future.

fduwjj replied:
Totally, which folder or directory do we feel is a better place?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1263,6 +1264,59 @@ def no_sync(self) -> Generator:
+        max_norm = float(max_norm)
+        norm_type = float(norm_type)
+        # Computes the max norm for this shard's gradients and sync's across workers
+        local_norm = _calc_grad_norm(self.params, norm_type).cuda()
```
why do we need to call `.cuda()` here? Seems like `_calc_grad_norm` should return the value on the same device as parameters in self.params?

fduwjj replied:
Right, but when the cpu_offload is set true, the params are on CPU. So looks like we do need this here otherwise, the test fails when calling the c10d API.

rohan-varma replied:
I see, makes sense. This implicitly assumes user has specified this rank's device appropriately with `torch.cuda.set_device()` which they should do, but could you add a small comment here anyways to give potential debugging hints to future devs :)

zhaojuanmao replied:
I think it is because backend of process group is NCCL, and all_reduce below expects CUDA tensors. yes, add a comment here will be great

fduwjj replied:
sure.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1263,6 +1264,59 @@ def no_sync(self) -> Generator:
+            dist.all_reduce(total_norm, group=self.process_group)
+            total_norm = total_norm ** (1.0 / norm_type)
+
+        if self.cpu_offload:
```
nice, let's also add a unittest for CPU offload that exercises this code path.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1263,6 +1264,59 @@ def no_sync(self) -> Generator:
+            total_norm = total_norm.cpu()
+
+        clip_coef = torch.tensor(max_norm, dtype=total_norm.dtype, device=total_norm.device) / (total_norm + 1e-6)
+        if clip_coef < 1:
```
what happens if clip_coef >= 1, is it expected to do nothing?

fduwjj replied:
Yep, because the norm is less than `max_norm`, so I guess this is the expected behavior?

rohan-varma reviewed (commented):
This is awesome work, thanks so much for persisting through all the usability issues and giving feedback on them offline! Added some comments for your consideration.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1263,6 +1264,59 @@ def no_sync(self) -> Generator:
                 m._require_backward_grad_sync = old_flag
 
+    @torch.no_grad()
+    def _clip_grad_norm_(
```
nit: _clip_grad_norm_ => clip_grad_norm_?

fduwjj replied:
Do you mean make it public/explicit? Like `_clip_grad_norm` => `clip_grad_norm_`.

zhaojuanmao replied:
yes, to be consistent with torch.nn.utils. clip_grad_norm_

fduwjj replied:
sure.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1263,6 +1264,59 @@ def no_sync(self) -> Generator:
+        .. warning:: This needs to be called on all ranks, since synchronization
+            primitives will be used.
+        """
+        self._lazy_init()
```
nit: add a comment to explain why _lazy_init() is needed

fduwjj replied:
sure.

zhaojuanmao reviewed (commented):
nice work! thanks for adding this. 

for tests, would you please add a norm_type config to _train_for_several_steps() in common_fsdp.py and compare ddp vs fsdp parity, similar to fairscale tests in _train_for_several_steps of test_fsdp.py for test_clip_norm_transformer and test_mixture_of_experts_grad_clip_breaks

Commit:
Update on "[PT-D][FSDP] Implement _clip_grad_norm_ for FSDP"


Implement the `_clip_grad_norm_` for FSDP, issue: https://github.com/pytorch/pytorch/issues/72548

Differential Revision: [D34230605](https://our.internmc.facebook.com/intern/diff/D34230605/)

[ghstack-poisoned]

Commit:
Update on "[PT-D][FSDP] Implement _clip_grad_norm_ for FSDP"


Implement the `_clip_grad_norm_` for FSDP, issue: https://github.com/pytorch/pytorch/issues/72548

Differential Revision: [D34230605](https://our.internmc.facebook.com/intern/diff/D34230605/)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -432,6 +440,11 @@ def _train_for_several_steps(
+                if isinstance(model, FullyShardedDataParallel):
+                    model.clip_grad_norm_(clip_norm, norm_type)
+                else:
+                    torch.nn.utils.clip_grad_norm_(model.parameters(), clip_norm, norm_type)
```
is this to test that DDP and FSDP have parity when clipping gradients? If so can you add some simple logic ensuring gradients were actually clipped (even something as simple as making sure total norm is under clip_norm value)

fduwjj replied:
sure.

rohan-varma commented on test/distributed/fsdp/test_fsdp_clip_grad_norm.py:
```diff
@@ -0,0 +1,136 @@
+        return total_norm ** (1.0 / norm_type)
+
+
+class TestClipGradNorm(FSDPTest):
```
Let's add a test for nested FSDP model, then overall it looks great!

fduwjj replied:
Sure.

Commit:
Update on "[PT-D][FSDP] Implement _clip_grad_norm_ for FSDP"


Implement the `_clip_grad_norm_` for FSDP, issue: https://github.com/pytorch/pytorch/issues/72548

Differential Revision: [D34230605](https://our.internmc.facebook.com/intern/diff/D34230605/)

[ghstack-poisoned]

fduwjj commented:
Adding the test coverage for the nested model.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1757,6 +1758,60 @@ def no_sync(self) -> Generator:
+        max_norm = float(max_norm)
+        norm_type = float(norm_type)
+        # Computes the max norm for this shard's gradients and sync's across workers
+        local_norm = _calc_grad_norm(self.parameters(), norm_type).cuda()  # type: ignore[arg-type]
```
could we replace self.parameters() with params_with_grad?

    @property
    def params_with_grad(self) -> List[Parameter]:
        """[p for p in self.parameters() if p.grad is not None]"""
        return [p for p in self.parameters() if p.grad is not None]

fduwjj replied:
sure, let me change that.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1757,6 +1758,60 @@ def no_sync(self) -> Generator:
+        clip_coef = torch.tensor(max_norm, dtype=total_norm.dtype, device=total_norm.device) / (total_norm + 1e-6)
+        if clip_coef < 1:
+            # multiply by clip_coef, aka, (max_norm/total_norm).
+            for p in self.parameters():
```
same here, use params_with_grad

zhaojuanmao reviewed (commented):
overall looks great to me, just one minor comment, will let @rohan-varma to accept it.

Commit:
Update on "[PT-D][FSDP] Implement _clip_grad_norm_ for FSDP"


Implement the `_clip_grad_norm_` for FSDP, issue: https://github.com/pytorch/pytorch/issues/72548

Differential Revision: [D34230605](https://our.internmc.facebook.com/intern/diff/D34230605/)

[ghstack-poisoned]

fduwjj commented:
Further address reviewer's comment.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1757,6 +1758,65 @@ def no_sync(self) -> Generator:
 
+    @property
+    def params_with_grad(self) -> List[Parameter]:
+        """[p for p in self.parameters() if p.grad is not None]"""
```
nit: comment is redundant because it is the next line itself. 

Maybe
"""
Recursively returns a list of all module parameters that have a gradient.
"""

rohan-varma reviewed (approved):
Awesome work, thanks for persisting through all of the issues and addressing all the comments! Looks great to ship.

Commit:
Update on "[PT-D][FSDP] Implement _clip_grad_norm_ for FSDP"


Implement the `_clip_grad_norm_` for FSDP, issue: https://github.com/pytorch/pytorch/issues/72548

Differential Revision: [D34230605](https://our.internmc.facebook.com/intern/diff/D34230605/)

[ghstack-poisoned]
